### ISSUE TITLE
Mutations and MOD_ACTIVATION bug?

### ISSUE THREAD

**Issue by tracycollins** [wagenaartje/neataptic#37]
Hi,

I'm new to neural networks and neataptic, so I assume that I'm misunderstanding something.

I'm trying to evolve a network with the following parameter for mutation:
 neataptic.Methods.Mutation.FFW

However, if I print out the js mutation object, I get nulls in the allowed MOD_ACTIVATION property :

 ```
 "mutation": [
    {
      "name": "ADD_NODE"
    },
    {
      "name": "SUB_NODE",
      "keep_gates": true
    },
    {
      "name": "ADD_CONN"
    },
    {
      "name": "REMOVE_CONN"
    },
    {
      "name": "MOD_WEIGHT",
      "min": -1,
      "max": 1
    },
    {
      "name": "MOD_BIAS",
      "min": -1,
      "max": 1
    },
    {
      "name": "MOD_ACTIVATION",
      "mutateOutput": true,
      "allowed": [
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        null
      ]
    },
    {
      "name": "SWAP_NODES",
      "mutateOutput": true
    }
  ]

```
Is this expected?  Am I doing something wrong?

I'm using the latest version of neataptic (as of today) v1.2.34

Thanks for any help.

**Comment by wagenaartje** [comment]
From what i'm seeing, you are converting an object (`Methods.Mutation`) to a JSON and _then_ printing it. Every element in the `MOD_ACTIVATION.allowed` array is a _function_, and functions **can't** be converted to JSON. That is why you are getting `null`.

If you run [this JSFiddle](https://jsfiddle.net/3rh6wu81/1/) you will notice what I mean. To print functions, use `toString()` instead.

So i'm closing this as this is expected behaviour of Javascript. If you have anymore questions, feel free to ask :+1:

**Comment by tracycollins** [comment]
Ah, yes. ok. Thanks for the quick reply, and thanks for this module.

I'm a neural network newbie, so I'm sure I'll have more questions.

One, unrelated question: Can you point me at examples (beyond the simple ones in this repository)? I'm trying to use evolution to generate a neural net that will categorize entities (using ~500 one-hot inputs) into one of 3 buckets (outputs). I'm not having much luck in generating a net that gets more than 50% success (barely better than random), and I'm not sure where I'm going wrong.

I suspect that my training set is too small (only about 1000), and that I probably need to run for more generations (1000 generations right now.)

Any help appreciated.

**Comment by wagenaartje** [comment]
I'm very busy with developing `1.3.0`, once it has been released i'll create some bigger examples (with the [MNIST dataset](https://github.com/cazala/mnist) for example). Could you show me what code you're using? Maybe I can give some tips.

**Comment by tracycollins** [comment]
Thanks for the offer to help. 

I suspect that I'm not choosing the best options for evolution and/or don't have enough data to adequately train and/or don't evolve the network for enough generations.

Here's my evolve function:

```
function evolve(params, callback){

  let options = {
    mutation: params.mutation, / /neataptic.Methods.Mutation.FFW
    cost: params.cost, // neataptic.Methods.Cost.MSE
    equal: params.equal, // false
    popsize: params.popsize, // 100
    elitism: params.elitism, // 5
    log: params.log, // 1
    iterations: params.iterations, // 100 - 1000, doesn't seem to matter in success rate
    mutationRate: params.mutationRate, // 0.7
    activation: params.activation // neataptic.Methods.Activation.LOGISTIC
  };

  // 1325 inputs, 3 outputs (1-hot or all 0 outputs)

  network = new neataptic.Network(numInputs, numOutputs); 

 // ~1000-item training set

  let trainingSet = [];

// each training set item:  datumObj = { name: "item name", datum: datum}

  async.each(params.trainingSet, function(datumObj, cb){

    debug("DATUM | " + datumObj.name);

    trainingSet.push(datumObj.datum);

    cb();

  }, function(){

    const results = network.evolve(trainingSet, options);

    if (callback !== undefined) { callback(results); }
  });
}

```
I rarely get more than 50% success rate (i.e., not better than random).

Any pointers greatly appreciated.

**Comment by wagenaartje** [comment]
Are you sure your processed trainingset is in the form of:

```javascript
var trainingSet = [
  { input: Array, output: Array },
  { input: Array, output: Array },
  { Array... 
];
```

Could you show me óne sample of your processed training set?

**Comment by tracycollins** [comment]
Here's a sample, with most of the input array deleted (because it's 1300+ inputs). The first 2 inputs are floating point numbers (normalized); the rest are binary.

I have an extra property called "inputHits" that I use for logging. Would this be a problem? I've also tried without the "inputHits" property, and it doesn't seem to make a difference.

`{
  "inputHits": 5,
  "input": [
    0.3488371938289734,
    0.20000000298023224,
    0,
    0
  ],
  "output": [
    1,
    0,
    0
  ]
}
`
